dashboard/app/jobs/aichat_request_chat_completion_job.rb

```diff
@@ -0,0 +1,80 @@
+class AichatRequestChatCompletionJob < ApplicationJob
+  queue_as :default
+
+  before_enqueue do |job|
+    request = job.arguments.first[:request]
+    request.update(execution_status: SharedConstants::AI_REQUEST_EXECUTION_STATUS[:QUEUED])
+  end
+
+  before_perform do |job|
+    request = job.arguments.first[:request]
+    request.update(execution_status: SharedConstants::AI_REQUEST_EXECUTION_STATUS[:RUNNING])
+  end
+
+  # Catch any exceptions that occur during the job and update the request status accordingly
+  rescue_from Exception do |exception|
+    if rack_env?(:development)
+      puts "AichatRequestChatCompletionJob Error: #{exception.full_message}"
+    end
+
+    request = arguments.first[:request]
+    request.update(response: exception.message, execution_status: SharedConstants::AI_REQUEST_EXECUTION_STATUS[:FAILURE])
+    Honeybadger.notify(
+      "AichatRequestChatCompletionJob failed with unexpected error: #{exception.message}",
+      context: {
+        request: request.to_json
+      }
+    )
+  end
+
+  def perform(request:, locale:)
+    model_customizations = JSON.parse(request.model_customizations, {symbolize_names: true})
+    stored_messages = JSON.parse(request.stored_messages, {symbolize_names: true})
+    new_message = JSON.parse(request.new_message, {symbolize_names: true})
+
+    status, response = get_execution_status_and_response(model_customizations, stored_messages, new_message, locale)
+    request.update(response: response, execution_status: status)
+  end
+
+  private def get_execution_status_and_response(model_customizations, stored_messages, new_message, locale)
+    # Check input for profanity and PII
+    user_profanity = find_profanity(new_message[:chatMessageText], locale)
+    return [SharedConstants::AI_REQUEST_EXECUTION_STATUS[:USER_PROFANITY], "Profanity detected in user input: #{user_profanity}"] if user_profanity
+
+    user_pii = find_pii(new_message[:chatMessageText], locale)
+    return [SharedConstants::AI_REQUEST_EXECUTION_STATUS[:USER_PII], "PII detected in user input: #{user_pii}"] if user_pii
+
+    # Make the request
+    response = AichatSagemakerHelper.get_sagemaker_assistant_response(model_customizations, stored_messages, new_message)
+
+    # Check output for profanity and PII. Report to HoneyBadger if the model returned profanity.
+    model_profanity = find_profanity(response, locale)
+    if model_profanity
+      Honeybadger.notify(
+        'Profanity returned from aichat model (blocked before reaching student)',
+        context: {
+          response: response,
+          flagged_content: model_profanity,
+        }
+      )
```

Comment:
sorry to potentially stir up a contentious topic, but if there's no code issue to be fixed in response to this then I would suggest not logging it to HoneyBadger. could it be logged to CloudWatch instead? If you're aware honeybadger isn't ideal, but none of the other alternatives meet your needs, then I think it is probably ok to keep logging there until firehose is fixed or replaced, but if that's the case it'd be nice to hear why HB is necessary in this case.

Reply:
Good question :) this is copied from what we're currently doing in the non-ActiveJob version of this API. Was previously discussed in this [PR](https://github.com/code-dot-org/code-dot-org/pull/58351/files#r1591315816) and this [thread](https://codedotorg.slack.com/archives/C046G4TRLEN/p1714672239480129). I think the decision more or less came down to, HB isn't ideal but we decided it logically made sense to treat profane responses as "errors" since that indicates that there's something wrong with our system prompt or other parameters. CloudWatch was considered as an alternative but as per these discussions it looks like there wasn't an easy way to implement it at the time?

Reply:
Thank you for the context! Honey Badger seems like the best option, given that you want to be notified promptly and take action.